modules/distribution-zones/src/test/java/org/apache/ignite/internal/distributionzones/DistributionZoneManagerAlterFilterTest.java

```diff
@@ -0,0 +1,340 @@
+        preparePrerequisites(10000, INFINITE_TIMER_VALUE);
+
+        assertNull(distributionZoneManager.zonesState().get(ZONE_ID).scaleUpTask());
+
+        topology.putNode(D);
+
+        // Check that scale up task was scheduled.
+        assertTrue(waitForCondition(() -> distributionZoneManager.zonesState().get(ZONE_ID).scaleUpTask() != null, 10_000));
+
+        String newFilter = "$[?(@.region == 'CN')]";
+
+        distributionZoneManager.alterZone(ZONE_NAME,
+                new DistributionZoneConfigurationParameters.Builder(ZONE_NAME)
+                        .filter(newFilter)
+                        .build()
+        ).get(10_000, TimeUnit.MILLISECONDS);
+
+        LogicalNode e = new LogicalNode(
+                new ClusterNode("5", "E", new NetworkAddress("localhost", 123)),
+                Map.of("region", "CN", "storage", "HDD", "dataRegionSize", "20")
+        );
+
+        doAnswer(invocation -> {
+            If iif = invocation.getArgument(0);
+
+            // Emulate a situation when immediate timer was run after filter altering and new node was added, so timer was scheduled.
+            byte[] key = zoneScaleUpChangeTriggerKey(ZONE_ID).bytes();
+
+            if (Arrays.stream(iif.cond().keys()).anyMatch(k -> Arrays.equals(key, k))) {
+                assertNotNull(distributionZoneManager.zonesState().get(1).scaleUpTask());
+
+                topology.putNode(e);
+            }
+            return invocation.callRealMethod();
+        }).when(keyValueStorage).invoke(any(), any());
+
+        // Check that node E, that was added while filter's altering, is not propagated to data nodes.
+        // TODO: https://issues.apache.org/jira/browse/IGNITE-19506 change this to the causality versioned call to dataNodes.
+        assertDataNodesFromManager(
+                distributionZoneManager,
+                ZONE_ID,
+                Set.of(C, D).stream().map(ClusterNode::name).collect(Collectors.toSet()),
+                10_000
+        );
+
+        // Assert that scheduled timer was not canceled because of immediate scale up after filter altering.
+        assertNotNull(distributionZoneManager.zonesState().get(ZONE_ID).scaleUpTask());
+
+        distributionZoneManager.alterZone(ZONE_NAME,
+                new DistributionZoneConfigurationParameters.Builder(ZONE_NAME)
+                        .dataNodesAutoAdjustScaleUp(IMMEDIATE_TIMER_VALUE)
+                        .dataNodesAutoAdjustScaleDown(IMMEDIATE_TIMER_VALUE)
+                        .filter(newFilter)
+                        .build()
+        ).get(10_000, TimeUnit.MILLISECONDS);
+
+        // Check that node E, that was added after filter's altering, was added only after altering immediate scale up.
+        assertDataNodesFromManager(
+                distributionZoneManager,
+                ZONE_ID,
+                Set.of(C, D, e).stream().map(ClusterNode::name).collect(Collectors.toSet()),
+                10_000
+        );
+    }
```

Comment:
Some tests are missing, e.g. altering from something-matches to nothing-matches.

Reply:
added